boxoffice/models/order.py

```diff
@@ -1,38 +1,240 @@
-from decimal import Decimal
-from boxoffice.models import db, BaseMixin
-from boxoffice.models.event import Event
-from boxoffice.models.user import User
-
-__all__ = ['Order', 'HALF_DOZEN']
-
-HALF_DOZEN = [
-    [0, "0"],
-    [1, "1"],
-    [2, "2"],
-    [3, "3"],
-    [4, "4"],
-    [5, "5"],
-    [6, "6"]
-]
+import random
+from datetime import datetime
+import decimal
+from collections import namedtuple
+from boxoffice.models import db, BaseMixin, User, Item, ItemCollection, Price
+from coaster.utils import LabeledEnum, buid
+from baseframe import __
+
+__all__ = ['Order', 'LineItem', 'PaymentTransaction']
+
+
+class ORDER_STATUS(LabeledEnum):
+    PURCHASE_ORDER = (0, __("Purchase Order"))
+    SALES_ORDER = (1, __("Sales Order"))
+    INVOICE = (2, __("Invoice"))
+    CANCELLED = (3, __("Cancelled Order"))
+
 
 class Order(BaseMixin, db.Model):
+    __tablename__ = 'customer_order'
+    __uuid_primary_key__ = True
+    __tableargs__ = (db.UniqueConstraint('item_collection_id', 'order_hash'),
+        db.UniqueConstraint('access_token'))
+
+    user_id = db.Column(None, db.ForeignKey('user.id'), nullable=True)
+    user = db.relationship(User, backref=db.backref('orders', cascade='all, delete-orphan'))
+    item_collection_id = db.Column(None, db.ForeignKey('item_collection.id'), nullable=False)
+    item_collection = db.relationship(ItemCollection, backref=db.backref('orders', cascade='all, delete-orphan'))  # noqa
+    status = db.Column(db.Integer, default=ORDER_STATUS.PURCHASE_ORDER, nullable=False)
+
+    initiated_at = db.Column(db.DateTime, nullable=False, default=datetime.utcnow)
+    invoiced_at = db.Column(db.DateTime, nullable=True)
+    cancelled_at = db.Column(db.DateTime, nullable=True)
+
+    access_token = db.Column(db.Unicode(22), nullable=False, default=buid)
+
+    buyer_email = db.Column(db.Unicode(254), nullable=False)
+    buyer_fullname = db.Column(db.Unicode(80), nullable=False)
+    buyer_phone = db.Column(db.Unicode(13), nullable=False)
```

Comment:
I did some research on phone number length and found 16 is the best size to accommodate all numbers in the world ([see here](https://github.com/hasgeek/lastuser/blob/db777be0ce2366a2620caef83973ddaf71305a50/lastuser_core/models/user.py#L855)).